pkg/hash/reader.go

```diff
@@ -165,3 +135,63 @@ func (r *Reader) Verify() error {
 	}
 	return nil
 }
+
+// merge another hash into this one.
+// There cannot be conflicting information given.
+func (r *Reader) merge(size int64, md5Hex, sha256Hex string, actualSize int64, strictCompat bool) (*Reader, error) {
+	if r.bytesRead > 0 {
+		return nil, errors.New("internal error: Already read from hash reader")
+	}
+	// Merge sizes.
+	// If not set before, just add it.
+	if r.size < 0 && size >= 0 {
+		r.src = io.LimitReader(r.src, size)
+		r.size = size
+	}
+	// If set before and set now they must match.
+	if r.size >= 0 && size >= 0 && r.size != size {
+		return nil, ErrSizeMismatch{Want: r.size, Got: size}
+	}
+
+	if r.actualSize <= 0 && actualSize >= 0 {
+		r.actualSize = actualSize
+	}
+
+	// Merge SHA256.
+	sha256sum, err := hex.DecodeString(sha256Hex)
+	if err != nil {
+		return nil, SHA256Mismatch{}
+	}
+
+	// If both are set, they must expect the same.
+	if r.sha256Hash != nil && len(sha256sum) > 0 {
+		if !bytes.Equal(r.sha256sum, sha256sum) {
+			return nil, SHA256Mismatch{}
+		}
+	} else if len(sha256sum) > 0 {
+		r.sha256Hash = sha256.New()
+		r.sha256sum = sha256sum
+	}
+
+	// Merge MD5 Sum.
+	md5sum, err := hex.DecodeString(md5Hex)
+	if err != nil {
+		return nil, BadDigest{}
+	}
+	// If both are set, they must expect the same.
+	if r.md5Hash != nil && len(md5sum) > 0 {
+		if !bytes.Equal(r.md5sum, md5sum) {
+			return nil, BadDigest{}
+		}
+	} else if len(md5sum) > 0 || (r.md5Hash == nil && strictCompat) {
+		r.md5Hash = md5.New()
+		r.md5sum = md5sum
+	}
+	return r, nil
+}
+
+// Close and release resources.
+func (r *Reader) Close() error {
```

Comment:
Should we not compute the MD5 and eventually SHA256 sum and compare to the reference values in this case?

Reply:
perhaps yes for a premature Close we can if we haven't hit the EOF yet. @klauspost ?

Reply:
@aead @harshavardhana (busy)

It may close for various reasons, so it may be fine for it to close, say if the request was aborted. Not the biggest fan of returning errors on Close since it is rarely checked anyway.

I considered adding a length check, so it would return an error if it gets an `io.EOF` *before* the expected number of bytes is read.

It would of course only trigger if the expected hash (if any) mismatches.

Furthermore it also doesn't limit on `size == 0`.

All of these 'features' of the existing implementation that would require vetting all users to not cause regressions.